engine/src/main/java/org/terasology/rendering/world/dag/ShadowMapNode.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright 2016 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.rendering.world.dag;
+
+
+import org.lwjgl.opengl.GL11;
+import org.terasology.config.RenderingConfig;
+import org.terasology.math.TeraMath;
+import org.terasology.math.geom.Vector3f;
+import org.terasology.rendering.assets.material.Material;
+import org.terasology.rendering.backdrop.BackdropProvider;
+import org.terasology.rendering.cameras.Camera;
+import org.terasology.rendering.cameras.OrthographicCamera;
+import org.terasology.rendering.opengl.FBO;
+import org.terasology.rendering.opengl.GraphicState;
+import org.terasology.rendering.primitives.ChunkMesh;
+import org.terasology.rendering.world.WorldRendererImpl;
+
+import static org.lwjgl.opengl.EXTFramebufferObject.GL_FRAMEBUFFER_EXT;
+import static org.lwjgl.opengl.EXTFramebufferObject.glBindFramebufferEXT;
+import static org.lwjgl.opengl.GL11.*;
+
+
+/**
+ * Diagram of this node can be viewed from:
+ * - https://docs.google.com/drawings/d/13I0GM9jDFlZv1vNrUPlQuBbaF86RPRNpVfn5q8Wj2lc/edit?usp=sharing
+ */
+public class ShadowMapNode implements RenderNode {
+    private static final int SHADOW_FRUSTUM_BOUNDS = 500;
+
+    private Material shader;
+    private FBO fbo;
+    private Camera camera = new OrthographicCamera(-SHADOW_FRUSTUM_BOUNDS, SHADOW_FRUSTUM_BOUNDS, SHADOW_FRUSTUM_BOUNDS, -SHADOW_FRUSTUM_BOUNDS);
+
+    // TODO: every node proposes its modifiable configuration?
+    private RenderingConfig renderingConfig;
+    private GraphicState state;
+
+    private Camera playerCamera;
+    // FIXME: remove reference to WorldRendererImpl
+    private WorldRendererImpl worldRenderer;
+    private BackdropProvider backdropProvider;
+
+
+    // FIXME: unnecessary arguments must be eliminated
+    public ShadowMapNode(Material shader,
+                         Camera playerCamera,
+                         WorldRendererImpl worldRenderer,
+                         BackdropProvider backdropProvider,
+                         RenderingConfig renderingConfig,
+                         GraphicState state) {
+
+        this.shader = shader;
```

Comment:
Hmmmm... I don't like the generic name. But I do not like the asset name either ("shadowMap"). I'm thinking "shadowCaster" or perhaps "shadowCastingShader"?